org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/services/EndpointProxy.java

```diff
@@ -68,6 +75,10 @@ public Object invoke(Object proxy, Method method, Object[] args) throws Throwabl
 		if (delegateInfo != null) {
 			return delegateInfo.delegate;
 		}
+		if (object_equals.equals(method) && args.length == 1 && args[0] == proxy) {
```

Comment:
Maybe delegate hashCode and equals on the proxy to the EndpointProxy itself? Proxy.getInvocationHandler will allow you to obtain the InvokationHandler of the proxy argument that is passed into invoke. This would be more symmetric, I suppose. Also a proper toString could be implemented that way.